README.md
<!-- PROJECT LOGO -->
<br />
<p align="center">
  <a href="https://github.com/elblogbruno/NotionAI-MyMind/">
    <img src="Chrome%20and%20Firefox%20Extension/icon/icon.png" alt="Logo" width="80" height="80">
  </a>
  

  
  <h3 align="center">NotionAI MyMind</h3>
  
  <p align="center">
    This repo uses AI and the wonderful Notion to enable you to add anything on the web to your "Mind" and forget about everything else.
    <br />
    <a href="https://github.com/elblogbruno/NotionAI-MyMind/wiki"><strong>Explore the docs »</strong></a>
    <br />
    <br />
    <a href="https://www.notion.so/glassear/My-Mind-Demo-Structure-ebd913d0cfa14889b122d391a883db94">View Demo</a>
    ·
    <a href="https://github.com/elblogbruno/NotionAI-MyMind/issues">Report Bug</a>
    ·
    <a href="https://github.com/elblogbruno/NotionAI-MyMind/discussions/categories/ideas">Request Feature</a>
  </p>
  <p align="center">
    <a href='https://chrome.google.com/webstore/detail/notion-ai-my-mind/eaheecglpekjjlegffodbfhbhdmnjaph?hl=es&authuser=0 '><img align="center" style="overflow: hidden; display: inline-block;" alt='Get it on Chrome Web Store' width=200          height=60 src='https://storage.googleapis.com/chrome-gcs-uploader.appspot.com/image/WlD8wC6g8khYWPJUsQceQkhXSlv1/HRs9MPufa1J1h5glNhut.png'/></a>  <a href='https://addons.mozilla.org/en-US/firefox/addon/notion-ai-my-mind/'><img align="center" style="overflow: hidden; display: inline-block;" alt='Get it on Firefox Add-On Store' width=172 height=60  src='https://ffp4g1ylyit3jdyti1hqcvtb-wpengine.netdna-ssl.com/addons/files/2015/11/get-the-addon.png'/></a>  <a href='https://play.google.com/store/apps/details?id=com.elblogbruno.notion_ai_my_mind&pcampaignid=pcampaignidMKT-Other-global-all-co-prtnr-py-PartBadge-Mar2515-1'><img align="center" style=" overflow: hidden; display: inline-block;" alt='Get it on Google Play' width=250 height=100 src='https://play.google.com/intl/en_us/badges/static/images/badges/en_badge_web_generic.png'/></a>
  
  </p>
  <a href="https://github.com/elblogbruno/NotionAI-MyMind/blob/master/readme-translations/README.es.md">
    <img src="https://img.shields.io/badge/lang-disponible%20en%20espa%C3%B1ol-brightgreen.svg" alt="Logo">
  </a>
</p>

### Add content 
<a href="https://github.com/elblogbruno/NotionAI-MyMind/wiki/Notion-AI-My-Mind-Collections"><strong>This is collections example, where you can have different collections or databases of contents, fully customizable on notion. »</strong></a>

<a href="https://github.com/elblogbruno/NotionAI-MyMind/wiki/Notion-AI-My-Mind-Collections">
    <img src="doc/header_collections_example_modify_title_tags.gif" alt="collections" width="690" height="388">
    <img src="doc/header_phone_demo.gif" alt="collections" height="388">
 </a>

### Search
<a href="https://www.notion.so/Intro-to-databases-fd8cd2d212f74c50954c11086d85997e"><strong>This is a fully customizable and searchable database in Notion. »</strong></a>

<a href="https://www.notion.so/Intro-to-databases-fd8cd2d212f74c50954c11086d85997e">
    <img src="doc/header_gif_search.gif" alt="collections" width="690" height="388">
 </a>
 
# Free your mind!

<!-- TABLE OF CONTENTS -->
<details open="open">
  <summary>Table of Contents</summary>
  <ol>
    <li>
      <a href="#project-philosophy">Project Philosophy</a>
      <ul>
        <li><a href="#examples-of-what-you-can-do">Examples</a></li>
        <li><a href="#extensions">Mind Extensions</a></li>
        <li><a href="#multilanguage-support-and-translation-collaboration">Translation into your language</a></li>
      </ul>
    </li>
    <li>
      <a href="#installing">Installing</a>
      <ul>
        <li><a href="#i-have-installed-the-server-what-to-do-next">I have installed the server, what to do next?</a></li>
        <li><a href="#docker-compose">Docker-Compose</a></li>
       </ul>
      </ul>
    </li>
    <li><a href="#common-issues">Common Issues</a></li>
    <li><a href="#roadmap">Roadmap</a></li>
  </ol>
</details>

## Project Philosophy.

The idea is to have extensions for your mind on the browser, and app on android and Ios, allowing you to add whatever you find on the web in your "Mind". Also, adding image and article tagging capabilities thanks to AI, so you can simply search on your "Mind" for what you remember.

Right now, there's a working Python Local Server, that receives all the data from the extension and the app, and publishes it to your fully customizable and searchable database in Notion. So it is 100% open source and fully private!

Maybe we can say it is an Open Source Alternative to [Raindrop](https://raindrop.io/) and [Microsoft Edge Collections](https://support.microsoft.com/en-us/microsoft-edge/organize-your-ideas-with-collections-in-microsoft-edge-60fd7bba-6cfd-00b9-3787-b197231b507e), but much cooler with Community driven opinion and AI Capabilities, and a repo maintainer with lot of imagination (yes my brain goes at 150% speed)!

## Examples of what you can do.

Add text to your mind  |  Add images to your mind
:---: | :---:
![](doc/add_text.gif)  |  ![](doc/add_image.gif)


Add websites to your mind |  Search on your mind
:---: | :---:
![](doc/add_website.gif)  |  ![](doc/header_gif_search.gif)

### Extensions
### Chromium users
The chromium based browsers such as Google Chrome,Brave or Microsoft edge among others, can install the extension from the store!

<a href='https://chrome.google.com/webstore/detail/notion-ai-my-mind/eaheecglpekjjlegffodbfhbhdmnjaph?hl=es&authuser=0 '><img alt='Get it on Chrome Web Store' width=200 height=60 src='https://storage.googleapis.com/chrome-gcs-uploader.appspot.com/image/WlD8wC6g8khYWPJUsQceQkhXSlv1/HRs9MPufa1J1h5glNhut.png'/></a>

### Firefox users
Firefox users can install the extension from the store!

<a href='https://addons.mozilla.org/en-US/firefox/addon/notion-ai-my-mind/'><img alt='Get it on Firefox Add-On Store' width=172 height=60 src='https://ffp4g1ylyit3jdyti1hqcvtb-wpengine.netdna-ssl.com/addons/files/2015/11/get-the-addon.png'/></a>

### Android and Ios users

Users can install the android app from android store. On Ios you can clone the flutter project and build the app.

<a href='https://play.google.com/store/apps/details?id=com.elblogbruno.notion_ai_my_mind&pcampaignid=pcampaignidMKT-Other-global-all-co-prtnr-py-PartBadge-Mar2515-1'><img alt='Get it on Google Play' width=250 height=100 src='https://play.google.com/intl/en_us/badges/static/images/badges/en_badge_web_generic.png'/></a>


I won't be releasing the app on the Apple App Store, as I don't have an Apple Developer Account either Mac OS based computer.
Meanwhile, you can clone the flutter project and build the app yourself.

## Multilanguage Support and Translation Collaboration

Since version 2.0.4 NotionAI-MyMind has multilanguage support! Now server, phone app and extension is translated to English and Spanish! 
Would you like to have it translated into your own language? 

You can have more info on how to collaborate it helps out people from all comunities and languages access this amazing tool! 
More info here:

https://github.com/elblogbruno/NotionAI-MyMind-Translations

## Installation Tutorial
[![Installation Tutorial](https://img.youtube.com/vi/v2wWtCYED1U/0.jpg)](https://www.youtube.com/watch?v=v2wWtCYED1U)

# Installing
- It is very easy, and there are different ways from click to install one's to more advanced ones, in case you want to install it from source.

- You can check it out on the wiki: [Installing the Notion AI My Mind Server](https://github.com/elblogbruno/NotionAI-MyMind/wiki/Installing-the-Notion-AI-My-Mind-Server)

- This covers:
    - Notion AI My Mind Server installation. 

### I have installed the server, what to do next?
- If you don't enter the Notion credentials either create the notion page, you would be not having fun with it!

- You can check it out on the wiki: [I have installed the server, what to do next?](https://github.com/elblogbruno/NotionAI-MyMind/wiki/I-have-installed-the-server,-what-to-do-next%3F)

- This covers:
    - Notion database creation
    - Browser or app walktrough with explanations.

- You can watch the video also on how to create the structure!
 
[![Installation Tutorial](https://img.youtube.com/vi/sRn6Pk1PnSY/0.jpg)](https://www.youtube.com/watch?v=sRn6Pk1PnSY)

### Docker-Compose

- You can check it out on the wiki: [Installing the Notion AI My Mind Server on Docker](https://github.com/elblogbruno/NotionAI-MyMind/wiki/Installing-the-Notion-AI-My-Mind-Server-on-Docker)

- This covers:
    - Server installation as a Docker Image

## Common Issues

- You can check it out on the wiki: [Common Issues](https://github.com/elblogbruno/NotionAI-MyMind/wiki/Common-Issues)

## Roadmap
- You can check the roadmap here: https://github.com/elblogbruno/NotionAI-MyMind/projects/1


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [2.0.6]

### Added 
- Now you can add a reminder to the content, so you get whatever you added reminded on the future whenever you want. If you select the Autodestroy option, booom , content will be auto-deleted from your mind at that specified time (Wish you had that in real life right?) pd: you can restore deleted notion content from the trash bin :wink:.
- Added a background worker that maintains the database, cleaning blank content added that fills the mind with garbage. (Check for duplicated content coming soon)
- Android app redesign as well, with bug fixes and less immersive experience.

### Fixed
- More error handling and structure fix to make everything faster and reliable.

## [2.0.5]

### Added 

- Now you can add Video files from web or phone!
- Started the base for adding a background worker that maintains the database, cleaning duplicated or blank content added. (Coming soon)
- Tags added from the AI tagging system have a number filter, so they are not all sent at the same time only the 10 first that are more used!

### Fixed
- More error handling and structure fix to make everything faster.
- Bug fix https://github.com/elblogbruno/NotionAI-MyMind/issues/20

## [2.0.4]

### Added 

- Now you can add Audio files from web or phone!
- Added multilanguage-support! Now server, phone app and extension is translated to English and Spanish! Would you like to have it translated into your own language? You can have more info on how to collaborate it helps out people from all comunities and languages access this amazing tool! More info [here](https://github.com/elblogbruno/NotionAI-MyMind-Translations).

- Chrome extension notifies you when there is a new update on the server!
- The server tells you now if there is a newer version of it on startup!
- Started the base for adding the ability to add notion reminders to content so it is reminded to you in the future. (Coming soon)
- Tags added from the AI tagging system are also suggested to you now when adding tags!

### Fixed
- Bug when first image was invalid, it did not add the next ones on the list.
- Bug that when adding content from phone, mind_extension property tell browser-extension instead of phone-extension
- More error handling
- Bug fix https://github.com/elblogbruno/NotionAI-MyMind/discussions/17
## [2.0.3]

### Breaking changes
- Now you can tell on the server the name of the properties, in case you want to customize the name of the basic properties that Notion AI My Mind contacts.

### Added 

- Now you can modify the title and url of the added content on the fly from the mind Extensions.
- You can add and remove multiple tags on the content you add. The tags you write will be saved to notion, so they will be suggested to you next time as you write.
  It also gets the current tags you may have added in notion for that property, and suggests them to you as you write!
- All this implements the ideas given by the community on #13 and #8.

- Added ability to scan QR Code on the android app, so you configure the app in no-time. When you access the server, you can scan it with the phone app.
- Fixed bugs and revamped the phone app UI. 
- Added more error handling on the phone app.
### Fixed

- More error handling

## [2.0.2]

### Breaking changes
- Notion API has changed. Maybe you get this error: "400 error on get_block() call",the fix is in here https://github.com/jamalex/notion-py/issues/292 but you need to modify manually the library. Meanwhile I am waiting for the library owner to update it.
- If you make a lot of request to notion API, means adding lot of content in a short amount of time, you may get now a 429 error(Too Many Requests). I may look into finding a solution or workaround to this, but it is a problem on notion's not my repo or @jamalex notion-py repo.

### Added 

- Added collections! Now you can have different collections or databases, so you can add the content to the collection you choose. More info here as you need to change your current notion setup:  https://github.com/elblogbruno/NotionAI-MyMind/wiki/Notion-AI-My-Mind-Collections

-Android app has updated to allow the new collections!

### Fixed

- Added more exceptions for error handling.
- Added notification for Notion's recent API change that limits api requests.

## [2.0.1]

### Added 

- Added function to modify a row title and url by its id, this is a base for developing ideas on #8.
- Now it looks at all images available on the content added so we can get more information and AI tag every image.
- Click to open added content works for url's, texts and images now.
- When you add content, it tracks if it was added from your phone or desktop. You need to add a new text atribute called "mind_extension". You can use this to filter by elements added from phone or desktop.
### Fixed

- Fixed error #9
- Fixed error no output provided when url or title was none adding a url.
- Fixed error when on some websites where tagged as image not found when images were found.
- Moved utils function to a util file.

## [2.0]

### Added 

- Added server as a .exe program on windows so it is easier as ever to run Notion AI My Mind in simple clicks.
- Server url is opened automatically on the browser, so you don't get lost if you don't know what the server url is.


### Fixed

- No port was returned at first run of the server.
- Refactored python directory, organized with folders.

## [1.9]

### Added 

- Faster and more reliable thanks to switching to an async based server with Quart. 
- Added json responses on server with built-in responses that would be translated to other languages. When requesting to the API, it returns the same text and status code on chrome, firefox and android/ios platforms.

Python-Server/app/server.py
import logging
from datetime import datetime
from urllib.error import HTTPError

import werkzeug
from quart import Quart, render_template, request, send_file
from werkzeug.utils import secure_filename

import secrets
import sys

from notion_ai.custom_errors import OnServerNotConfigured
from server_utils.handle_options_data import process_formulary
from notion_ai.notion_ai import *
from notion.client import NotionClient
from server_utils.utils import ask_server_port, createFolder, allowed_file, is_a_sound_file, is_a_video_file, \
    SETTINGS_FOLDER, UPLOAD_FOLDER
from server_utils.check_update import *
from notion_ai.utils import create_json_response
import tempfile

dir_path = 'notion-ai-app.log'
if getattr(sys, 'frozen', False):
    template_folder = os.path.join(sys._MEIPASS, 'templates')
    static_folder = os.path.join(sys._MEIPASS, 'static')
    tmp = tempfile.mkdtemp()
    dir_path = os.path.join(tmp, 'notion-ai-app.log')
    app_quart = Quart(__name__, template_folder=template_folder, static_folder=static_folder)
else:
    app_quart = Quart(__name__)

app_quart.config['UPLOAD_FOLDER'] = UPLOAD_FOLDER

logging.basicConfig(filename=dir_path, filemode='w', format='%(name)s - %(levelname)s - %(message)s',
                    level=logging.INFO)

# define a Handler which writes INFO messages or higher to the sys.stderr
console = logging.StreamHandler()
console.setLevel(logging.INFO)
# add the handler to the root logger
logging.getLogger('').addHandler(console)
logging.info("Log file will be saved to temporary path: {0}".format(dir_path))

notion_ai = None


@app_quart.route('/add_url_to_mind')
async def add_url_to_mind():
    try:
        url = request.args.get('url')
        title = request.args.get('title')
        collection_index = request.args.get('collection_index') if request.args.get('collection_index') else 0
        notion_ai.set_mind_extension(request)
        return str(notion_ai.add_url_to_database(url, title, int(collection_index), False))
    except OnServerNotConfigured as e:
        logging.error(e)
        return str(create_json_response(notion_ai, status_code=e.status_code, port=port))


@app_quart.route('/add_text_to_mind')
async def add_text_to_mind():
    try:
        url = request.args.get('url')
        text = request.args.get('text')
        collection_index = request.args.get('collection_index') if request.args.get('collection_index') else 0
        notion_ai.set_mind_extension(request)

        return str(notion_ai.add_content_to_database(content_src=text, content_type='text', url=url,
                                                     collection_index=int(collection_index)))
    except OnServerNotConfigured as e:
        logging.error(e)
        return str(create_json_response(notion_ai, status_code=e.status_code, port=port))


@app_quart.route('/add_image_to_mind')
async def add_image_to_mind():
    try:
        url = request.args.get('url')
        image_src = request.args.get('image_src')
        image_src_url = request.args.get('image_src_url')
        collection_index = request.args.get('collection_index') if request.args.get('collection_index') else 0

        notion_ai.set_mind_extension(request)

        return str(notion_ai.add_content_to_database(content_src=image_src, content_type='image', url=url,
                                                     image_src_url=image_src_url,
                                                     collection_index=int(collection_index)))
    except OnServerNotConfigured as e:
        logging.error(e)
        return str(create_json_response(notion_ai, status_code=e.status_code, port=port))


@app_quart.route('/add_audio_to_mind')
async def add_audio_to_mind():
    try:
        url = request.args.get('url')
        audio_src = request.args.get('audio_src')
        collection_index = request.args.get('collection_index') if request.args.get('collection_index') else 0

        notion_ai.set_mind_extension(request)
        return str(notion_ai.add_content_to_database(content_src=audio_src, content_type='audio', url=url,
                                                     collection_index=int(collection_index)))
    except OnServerNotConfigured as e:
        logging.error(e)
        return str(create_json_response(notion_ai, status_code=e.status_code, port=port))


@app_quart.route('/add_video_to_mind')
async def add_video_to_mind():
    try:
        url = request.args.get('url')
        video_src = request.args.get('video_src')
        collection_index = request.args.get('collection_index') if request.args.get('collection_index') else 0

        notion_ai.set_mind_extension(request)
        return str(notion_ai.add_content_to_database(content_src=video_src, content_type='video', url=url,
                                                     collection_index=int(collection_index)))
    except OnServerNotConfigured as e:
        logging.error(e)
        return str(create_json_response(notion_ai, status_code=e.status_code, port=port))


@app_quart.route('/get_mind_structure')
async def get_mind_structure():
    try:
        return str(notion_ai.mind_structure.get_mind_structure(notion_ai))
    except OnServerNotConfigured as e:
        logging.error("Get mind structure " + str(e))
        return str(create_json_response(notion_ai, port=port))
    except AttributeError as e:
        logging.error("Get mind structure attribute error " + str(e))
        return str(create_json_response(notion_ai, error_sentence=str(e), port=port))


def toDate(dateString):
    return datetime.strptime(dateString, "%Y-%m-%dT%H:%M") #example datetime input 2021-07-01T13:45



@app_quart.route('/set_reminder_date_to_block')
async def set_reminder_date_to_block():
    id = request.args.get("id")

    # if there is a week beetwen start and end 1 week before option appears. ({'time': '09:00', 'unit': 'week', 'value': 1})
    try:
        start = request.args.get('start', type=toDate)

        logging.info("Start date: " + str(start))
        # end = request.args.get('end', default=datetime.today(), type=toDate)
        unit = request.args.get('unit', default='minute')  # if unit is minute and value is 0 it means At Time of Event
        auto_destroy = request.args.get('autodestroy',
                                        default=False)  # if unit is minute and value is 0 it means At Time of Event
        remind_value = request.args.get('remind_value',
                                        default=0)  # it can be 5 minutes 10 15 or 30.  Aswell as 1 2 hours or 1 2 days before

        return str(notion_ai.set_reminder_date_to_block(logging, id=id, start=start, unit=unit,
                                                        remind_value=remind_value, self_destruction=auto_destroy))

    except ValueError as e:
        logging.error("Error parsing reminder start date" + str(e))
        return str(create_json_response(notion_ai, error_sentence=str(e), status_code=305))
    except HTTPError as e:
        logging.error("Error parsing reminder start date" + str(e))
        return str(create_json_response(notion_ai, error_sentence=str(e), status_code=305))



@app_quart.route('/modify_element_by_id')
async def modify_element_by_id():
    id = request.args.get('id')
    title = request.args.get('new_title')
    url = request.args.get('new_url')
    notion_ai.set_mind_extension(request)
    return str(notion_ai.modify_row_by_id(id, title, url))


@app_quart.route('/get_multi_select_tags')
async def get_multi_select_tags():
    try:
        collection_index = request.args.get('collection_index') if request.args.get('collection_index') else 0
        notion_ai.set_mind_extension(request)
        ai_tags = notion_ai.image_tagger.get_most_used_ai_tags(notion_ai=notion_ai,
                                                               collection_index=int(collection_index))
        return str(notion_ai.property_manager.multi_tag_manager.get_multi_select_tags(notion_ai,
                                                                                      append_tags=ai_tags))  # multi_select_tag_list
    except ValueError as e:
        logging.error(e)
        return str(create_json_response(notion_ai, error_sentence=str(e), status_code=401))
    except OnServerNotConfigured as e:
        logging.error(e)
        return str(create_json_response(notion_ai, status_code=e.status_code, port=port))


@app_quart.route('/update_multi_select_tags', methods=['POST'])
async def update_multi_select_tags():
    try:
        collection_index = request.headers["collection_index"]
        id = request.headers["id"]
        tags = await request.get_json()
        notion_ai.set_mind_extension(request)
        return str(
            notion_ai.property_manager.multi_tag_manager.update_multi_select_tags(notion_ai, id, tags,
                                                                                  collection_index))
    except ValueError as e:
        logging.error(e)
        return str(create_json_response(notion_ai, error_sentence=str(e), status_code=401))


@app_quart.route('/upload_file', methods=['POST'])
async def upload_file():
    try:
        status_code = 200
        # check if the post request has the file part
        request_files = await request.files

        notion_ai.set_mind_extension(request)

        collection_index = request.headers["collection_index"]

        if 'file' not in request_files:
            print('No file part')
            status_code = 400
            return str(create_json_response(notion_ai, status_code=status_code))

        file = request_files['file']
        # if user does not select file, browser also
        # submit an empty part without filename
        if file.filename == '':
            print('No selected file')
            status_code = 400
            return str(create_json_response(notion_ai, status_code=status_code))

        print(file.filename)

        if file and allowed_file(file.filename):
            filename = secure_filename(file.filename)
            file.save(os.path.join(app_quart.config['UPLOAD_FOLDER'], filename))
            uri = os.path.join(app_quart.config['UPLOAD_FOLDER'], filename)

            if is_a_sound_file(filename):
                return str(notion_ai.add_content_to_database(content_src=uri, content_type='audio', collection_index= int(collection_index)))
            elif is_a_video_file(filename):
                return str(notion_ai.add_content_to_database(content_src=uri, content_type='video', collection_index= int(collection_index)))
            else:
                return str(notion_ai.add_content_to_database(content_src=uri, content_type='image', collection_index= int(collection_index)))
        else:
            print("This file is not allowed to be post")
            status_code = 400
            return str(create_json_response(notion_ai, status_code=status_code))
    except werkzeug.exceptions.BadRequestKeyError as e:
        logging.error(e)
        return str(create_json_response(notion_ai, error_sentence=str(e), port=port, status_code=status_code))
    except OnServerNotConfigured as e:
        logging.error(e)
        return str(create_json_response(notion_ai, status_code=e.status_code, port=port))


@app_quart.route('/get_current_mind_url')
async def get_current_mind_url():
    try:
        return str(notion_ai.data['url'])
    except AttributeError as e:
        logging.error("get_current_mind_url atribute error" + str(e))
        return str(create_json_response(notion_ai, error_sentence=str(e), port=port))


@app_quart.route('/check_update')
async def check_update_api():
    return str(check_update(logging, app_quart.static_folder, notion_ai, port, return_response=True))


@app_quart.route('/log')
async def return_log_file():
    """Returns the log file of the server"""
    try:
        return await send_file(dir_path, as_attachment=True, attachment_filename='notion-ai-app.log')
    except Exception as e:
        return str(e)


@app_quart.route('/update_notion_tokenv2')
async def update_notion_tokenv2():
    token_from_extension = request.args.get('tokenv2')
    changed = False
    with open('data.json') as json_file:
        options = json.load(json_file)

        if token_from_extension != options['token']:
            try:
                options['token'] = token_from_extension

                client = NotionClient(token_v2=options['token'])  # if can't make a client out of the token, it is not
                # a correct one.

                a_file = open("data.json", "w")
                json.dump(options, a_file)
                a_file.close()

                logging.info("Token v2 changed to {}".format(token_from_extension))
                changed = notion_ai.run()
                if changed:
                    return create_json_response(notion_ai, status_code=2)
                else:
                    return create_json_response(notion_ai, status_code=3)
            except requests.exceptions.HTTPError as e:
                logging.error("Incorrect token V2 from notion")
                return create_json_response(notion_ai, error_sentence=str(e), status_code=3)


@app_quart.route('/')
async def show_settings_home_menu():
    return await render_template("options.html")


@app_quart.route('/handle_data', methods=['POST'])
async def handle_data():
    data = await request.get_json()

    process_formulary(logging, data)

    use_email = data['email'] and data['password']

    if use_email:
        has_run = notion_ai.run(logging, email=data['email'], password=data['password'])
    else:
        has_run = notion_ai.run(logging)

    if has_run:
        return "200"
    else:
        return "500"


if __name__ == "__main__":
    secret = secrets.token_urlsafe(32)
    app_quart.secret_key = secret
    createFolder("settings")
    createFolder("uploads")

    check_update(logging, app_quart.static_folder)
    port = ask_server_port(logging)
    notion_ai = NotionAI(logging, port, app_quart.static_folder)
    app_quart.run(host="0.0.0.0", port=port, debug=True)
